src/models/bgp/attributes/mod.rs

```diff
@@ -109,15 +122,296 @@ pub fn get_deprecated_attr_type(attr_type: u8) -> Option<&'static str> {
     }
 }
 
+/// Convenience wrapper for a list of attributes
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Attributes {
+    // Black box type to allow for later changes/optimizations. The most common attributes could be
+    // added as fields to allow for easier lookup.
+    inner: Vec<Attribute>,
+}
+
+impl Attributes {
+    pub fn has_attr(&self, ty: AttrType) -> bool {
+        self.inner.iter().any(|x| x.value.attr_type() == ty)
+    }
+
+    pub fn get_attr(&self, ty: AttrType) -> Option<Attribute> {
+        self.inner
+            .iter()
+            .find(|x| x.value.attr_type() == ty)
+            .cloned()
+    }
+
+    pub fn origin(&self) -> Origin {
+        self.inner
+            .iter()
+            .find_map(|x| match &x.value {
+                AttributeValue::Origin(x) => Some(*x),
+                _ => None,
+            })
+            .unwrap_or(Origin::INCOMPLETE)
+    }
+
+    pub fn origin_id(&self) -> Option<IpAddr> {
+        self.inner.iter().find_map(|x| match &x.value {
+            AttributeValue::OriginatorId(x) => Some(*x),
+            _ => None,
+        })
+    }
+
+    // TODO: Why would the attribute be used instead of the reach NLRI? Should this check both?
+    pub fn next_hop(&self) -> Option<IpAddr> {
```

Comment:
Would it be preferable to search for an `MP_REACHABLE_NLRI` attribute to get the next hop if the `NEXT_HOP` attribute is not found? From what I have seen, `NEXT_HOP` will sometimes not be included when `MP_REACHABLE_NLRI` is present despite being a mandatory attribute.

Reply:
I'd prefer for now to leave it as `next_hop` is an explicitly defined attribute and I don't want to do the smart thing and fill in the blank if it was mistakenly left empty. Unless there is a convincing legitimate case where `next_hop` should be added to the NLRI (or per-prefix next-hop in the same MRT record), I think we should just treat this as is.

Reply:
Sounds good.

For context though, I was curious about how frequently different attributes show up so I ran a quick test. In the test, I found that ~15% of `Attributes` within rib dumps did not contain `NEXT_HOP`.
```text
RIB Attribute Groups
COUNT         PERCENT NAME
[truncated]
77209439     1.40470% [ORIGIN, AS_PATH, NEXT_HOP, ATOMIC_AGGREGATE, COMMUNITIES]
82602566     1.50282% [ORIGIN, AS_PATH, NEXT_HOP, AGGREGATOR, COMMUNITIES]
83295574     1.51542% [ORIGIN, AS_PATH, NEXT_HOP, MULTI_EXIT_DISCRIMINATOR, COMMUNITIES, LARGE_COMMUNITIES]
96038091     1.74725% [ORIGIN, AS_PATH, NEXT_HOP, ATOMIC_AGGREGATE, AGGREGATOR, COMMUNITIES]
134907217    2.45441% [ORIGIN, AS_PATH, MULTI_EXIT_DISCRIMINATOR, COMMUNITIES, MP_REACHABLE_NLRI]
173909949    3.16400% [ORIGIN, AS_PATH, NEXT_HOP, MULTI_EXIT_DISCRIMINATOR]
212989898    3.87500% [ORIGIN, AS_PATH, MP_REACHABLE_NLRI]
241105973    4.38652% [ORIGIN, AS_PATH, NEXT_HOP, COMMUNITIES, LARGE_COMMUNITIES]
256900824    4.67388% [ORIGIN, AS_PATH, COMMUNITIES, MP_REACHABLE_NLRI]
737080900   13.40996% [ORIGIN, AS_PATH, NEXT_HOP, MULTI_EXIT_DISCRIMINATOR, COMMUNITIES]
1173907989  21.35731% [ORIGIN, AS_PATH, NEXT_HOP]
1564279355  28.45947% [ORIGIN, AS_PATH, NEXT_HOP, COMMUNITIES]

RIB Individual Attribute Occurrences
COUNT         PERCENT NAME
0            0.00000% AS4_PATH
0            0.00000% AIGP
0            0.00000% BGP_LS_ATTRIBUTE
0            0.00000% SFP_ATTRIBUTE
0            0.00000% PE_DISTINGUISHER_LABELS
0            0.00000% PMSI_TUNNEL
0            0.00000% BGPSEC_PATH
0            0.00000% BFD_DISCRIMINATOR
0            0.00000% TUNNEL_ENCAPSULATION
0            0.00000% TRAFFIC_ENGINEERING
0            0.00000% CLUSTER_ID
0            0.00000% ATTR_SET
0            0.00000% AS4_AGGREGATOR
0            0.00000% IPV6_ADDRESS_SPECIFIC_EXTENDED_COMMUNITIES
0            0.00000% MP_UNREACHABLE_NLRI
0            0.00000% BGP_PREFIX_SID
844          0.00002% Unknown(36)
1432         0.00003% Unknown(20)
1666         0.00003% Unknown(243)
5327         0.00010% Unknown(21)
405735       0.00738% DEVELOPMENT
838612       0.01526% ORIGINATOR_ID
838612       0.01526% CLUSTER_LIST
1043763      0.01899% LOCAL_PREFERENCE
17288435     0.31453% ONLY_TO_CUSTOMER
59600340     1.08433% EXTENDED_COMMUNITIES
446800968    8.12880% ATOMIC_AGGREGATE
523097119    9.51688% AGGREGATOR
591410005   10.75972% LARGE_COMMUNITIES
816150405   14.84850% MP_REACHABLE_NLRI
1336463411  24.31473% MULTI_EXIT_DISCRIMINATOR
3632860821  66.09388% COMMUNITIES
4680366060  85.15150% NEXT_HOP
5496516465 100.00000% AS_PATH
5496516465 100.00000% ORIGIN
```
I did find it strange though how `AS4_PATH` never occurred in the RIB dumps. I'm not sure if this indicates that something was broken during my changes.

The code and full data from that test is in this repo: https://github.com/jmeggitt/bgp_attribute_survey

Reply:
I think it's because `As4Path` is a value type that we created but not really used. In the parser function `parse_as_path`, we parse AS paths based on `asn_len` (i.e. 2byte or 4 byte) but will create a `AsPath` attribute value regardless of the length. I think we could probably remove the `As4Path` in the future.

Great work on the attribute survey! It is very interesting to learn about what attributes show up in the MRT files.